### PR TITLE
Rolling re-create support for Istiod and MGW deployments

### PR DESCRIPTION
**Pull request by waynz0r**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | yes
| API breaks?     | no
| Deprecations?   | no
| Related tickets | 
| License         | Apache 2.0

### What's in this PR?

Implements support for rolling upgrade like scenario for `immutable field change -> re-create` updates of Istiod and MeshGateway deployments.

### Why?

Immutable fields of a k8s resource cannot be changed with a resource update, in that case the operator removes the existing resource and creates a new one. For a deployment that will cause service disruption since the existing pods will be terminated before the new ones come up and become ready.

### Checklist

- [x] Implementation tested

### CODE REVIEW

File: pkg/k8sutil/desired_state.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	return s.afterCreateFunc(current, desired)
+}
+
+func (s RecreateAwareDesiredState) BeforeRecreate(current, desired runtime.Object) error {
+	return s.beforeCreateFunc(current, desired)
+}
+
+func DeploymentDesiredStateWithReCreateHandling(c client.Client, scheme *runtime.Scheme, log logr.Logger, podLabels map[string]string) RecreateAwareDesiredState {
+	podLabels = util.MergeMultipleStringMaps(map[string]string{
+		"detached": "true",
```
Comment (khernyo) [inline pkg/k8sutil/desired_state.go:76]: the label should be "istio.banzaicloud.io/detached" or similar.

**Review by khernyo — APPROVED** [review]

File: pkg/k8sutil/desired_state.go (hunk 1 above)
Comment (waynz0r) [reply pkg/k8sutil/desired_state.go]: 👍

File: pkg/k8sutil/detach_pod.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	err = c.List(context.Background(), &client.ListOptions{
+		LabelSelector: ls,
+	}, pods)
+	if err != nil {
+		return err
+	}
+
+	for _, pod := range pods.Items {
+		if len(pod.OwnerReferences) != 1 {
+			log.V(1).Info("evaluting pod for detaching", "action", "skip", "deploymentName", deployment.Name, "name", pod.Name, "reason", "notExactlyOneOwnerReference")
```
Comment (Laci21) [inline pkg/k8sutil/detach_pod.go:46]:
```suggestion
			log.V(1).Info("evaluting pod for detaching", "action", "skip", "deploymentName", deployment.Name, "name", pod.Name, "reason", "notExactlyOneOwnerReferenceForPod")
```

File: pkg/k8sutil/detach_pod.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		err = c.Get(context.Background(), client.ObjectKey{
+			Name:      ownerRef.Name,
+			Namespace: pod.Namespace,
+		}, rs)
+		if err != nil {
+			return err
+		}
+
+		if len(rs.OwnerReferences) != 1 {
+			log.V(1).Info("evaluting pod for detaching", "action", "skip", "deploymentName", deployment.Name, "name", pod.Name, "reason", "replicaSetHasMultipleOwners")
```
Comment (Laci21) [inline pkg/k8sutil/detach_pod.go:64]:
```suggestion
			log.V(1).Info("evaluting pod for detaching", "action", "skip", "deploymentName", deployment.Name, "name", pod.Name, "reason", "notExactlyOneOwnerReferenceForReplicaSet")
```

File: pkg/k8sutil/desired_state.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+)
+
+type DesiredState interface {
+	AfterRecreate(current, desired runtime.Object) error
+	BeforeRecreate(current, desired runtime.Object) error
+}
+
+type StaticDesiredState string
+
+func (s StaticDesiredState) AfterRecreate(current, desired runtime.Object) error {
```
Comment (Laci21) [inline pkg/k8sutil/desired_state.go:44]:
```suggestion
func (s StaticDesiredState) AfterRecreate(_, _ runtime.Object) error {
```

File: pkg/k8sutil/desired_state.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+	return nil
+}
+
+func (s StaticDesiredState) BeforeRecreate(current, desired runtime.Object) error {
```
Comment (Laci21) [inline pkg/k8sutil/desired_state.go:48]:
```suggestion
func (s StaticDesiredState) BeforeRecreate(_, _ runtime.Object) error {
```

**Review by Laci21 — APPROVED** [review]
lgtm, just a few nits

**Pull request merged**